src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,394 @@
+                }
+            }
+
+            combined = combinedBuilder.ToImmutable();
+            QsCompilerError.Verify(combined.Keys.All(key => key.Item1.Equals(parent)), "for type parameter that does not belong to parent callable");
+            return success;
+        }
+
+        /// <summary>
+        /// This is a dictionary mapping source nodes to information about target nodes. This information is represented
+        /// by a dictionary mapping target node to the edges pointing from the source node to the target node.
+        /// </summary>
+        private readonly Dictionary<CallGraphNode, Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>> _Dependencies =
+            new Dictionary<CallGraphNode, Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>>();
+
+        private QsNullable<ImmutableArray<ResolvedType>> RemovePositionFromTypeArgs(QsNullable<ImmutableArray<ResolvedType>> tArgs) =>
+            tArgs.IsValue
+            ? QsNullable<ImmutableArray<ResolvedType>>.NewValue(tArgs.Item.Select(StripPositionInfo.Apply).ToImmutableArray())
+            : tArgs;
+
+        private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey, CallGraphEdge edge)
+        {
+            if (_Dependencies.TryGetValue(callerKey, out var deps))
+            {
+                deps[calledKey] = deps.TryGetValue(calledKey, out var edges)
+                    ? edges.Add(edge)
+                    : ImmutableArray.Create(edge);
+            }
+            else
+            {
+                var newDeps = new Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>();
+                newDeps[calledKey] = ImmutableArray.Create(edge);
+                _Dependencies[callerKey] = newDeps;
+            }
+        }
+
+        /// <summary>
+        /// Adds a dependency to the call graph using the caller's specialization and the called specialization's information.
+        /// </summary>
+        public void AddDependency(QsSpecialization callerSpec, QsQualifiedName calledName, QsSpecializationKind calledKind, QsNullable<ImmutableArray<ResolvedType>> calledTypeArgs, CallGraphEdge edge) =>
+            AddDependency(
+                callerSpec.Parent, callerSpec.Kind, callerSpec.TypeArguments,
+                calledName, calledKind, calledTypeArgs,
+                edge);
+
+        /// <summary>
+        /// Adds a dependency to the call graph using the relevant information from the caller's specialization and the called specialization.
+        /// </summary>
+        public void AddDependency(
+            QsQualifiedName callerName, QsSpecializationKind callerKind, QsNullable<ImmutableArray<ResolvedType>> callerTypeArgs,
+            QsQualifiedName calledName, QsSpecializationKind calledKind, QsNullable<ImmutableArray<ResolvedType>> calledTypeArgs,
+            CallGraphEdge edge)
+        {
+            // ToDo: Setting TypeArgs to Null because the type specialization is not implemented yet
```

Comment:
Would you mind using the specific phrasing below sucht that it's easy to find these places in the code?
```suggestion
            // Setting TypeArgs to Null because the type specialization is not implemented yet
            // ToDo: this needs adaption if we want to support type specializations
```